### PR TITLE
Navigation Menubar Example: Improve high contrast support and refactor javascript

### LINKED ISSUE #1357: navigation menubar example: Improve high contrast support and refactor  Javascript
Update the navigation menubar Javascript to:

1. Javascript uses a single object
2. Improve support for high contrast mode

### PR DESCRIPTION
**Pull request by jongund**
closes #1357

* Updated the javascript to use a single object
* Updated example to improve high contrast support

#### Preview Link

[View navigation menubar in the compare branch for this PR](https://raw.githack.com/w3c/aria-practices/issue1357-navigation-menubar-js/examples/menubar/menubar-navigation.html)

#### Review Checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @a11ydoer
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @charmarkk 
* [x] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @carmacleod 
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @spectranaut

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/pull/1359.html" title="Last updated on Jul 19, 2020, 12:14 AM UTC (15e3450)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/1359/afafad7...15e3450.html" title="Last updated on Jul 19, 2020, 12:14 AM UTC (15e3450)">Diff</a>

### CODE REVIEW

**Comment by jongund** [comment]
@charmarkk 

Can you re-review the navigation menubar example for high contrast, I added a 1 pixel border to the menubar for this example too.

**Comment by a11ydoer** [comment]
@carmacleod @zcorpan @mcking65 
Could you kindly allot time to review the menubar example before APG meeting tomorrow so that we can publish the example? Thanks for your consideration.

**Comment by jongund** [comment]
@carmacleod 

Updated mouse hover behavior to match menubar editor changes based on teleconference earlier this week.

**Comment by smhigley** [comment]
@jongund fixed conflicts & updated tests with `t.plan` changes

**Comment by jongund** [comment]
@smhigley 
Thank you for updating the regression tests.

**Comment by jongund** [comment]
@spectranaut 

I removed the original menubar navigation example and test from the pull request.

**Comment by mcking65** [comment]
@jongund wrote:
> I removed the original menubar navigation example and test from the pull request.
> It doesn't seem to be giving the merge conflict for the removed test file this time.

Great, but confusing, news. I'd love to wrap this one up next then.

Jon, please submit a PR to delete the old editor menubar example and tests.

**Comment by carmacleod** [comment]
I've submitted a PR to update the "Similar examples" links on both Navigation and Menubar Example pages.
  - on Navigation Menubar page, changed to href="menubar-editor.html"
  - on Editor Menubar page, changed to href="menubar-navigation.html"

**Comment by jongund** [comment]
@carmacleod 
@mcking65 

I fixed the focus on hover issue for the menubar and updated the cross referencing links for menubar editor and menubar navigation.
If the focus on hover fix looks good for the menubar navigation pull request I will fix it in the menubar editor example in a separate pull request.

**Comment by carmacleod** [comment]
@jongund 
> I fixed the focus on hover issue for the menubar and updated the cross referencing links for menubar editor and menubar navigation.

Looks good! Thanks for the super-fast fix!

> If the focus on hover fix looks good for the menubar navigation pull request I will fix it in the menubar editor example in a separate pull request.

The problem doesn't happen in the Editor Menubar Example, so no need to fix that one. :)

**Comment by jongund** [comment]
@carmacleod 
@mcking65 
Carolyn thank you for all your feedback and comments.
I have fixed the text selection issue, was canceling mouse event propagation.
Please retest.

**Comment by carmacleod** [comment]
@jongund The text selection issue is fixed - thank-you!

**Comment by carmacleod** [comment]
I tested both the Navigation and Editor Menubar Examples in Windows High Contrast Mode on Firefox, Edge, and IE.
(Chrome still doesn't have HCM available in the release, and I threw in IE to make sure there weren't any disappearing background images - there weren't any, and since IE has problems with svg sizing, I won't discuss IE any further).

Unfortunately, we haven't finished [documenting our color contrast testing procedure](https://github.com/w3c/aria-practices/issues/1392) so I decided to test High Contrast Black and High Contrast White. Here's a screen snap of both sets of colors. (Sorry that a lot of this comment will be screen snapshots).

High Contrast Black and High Contrast White colors

### High Contrast Black in Firefox

Checkmarks, radio dots, and separators are blue, which unfortunately only has a 2.1:1 contrast ratio on black background. They are white when focused, which is good on black. Drop-down triangles are white when expanded (good), but the blue and the grey colors don't pass contrast. Firefox colors visited links a different color, which looks a bit odd in a menu, but has good contrast.

image

### High Contrast Black in Edge

The Editor Menubar Example in Edge is the same as Firefox, but the Navigation Menubar has one slight difference because Edge uses one color for both visited and unvisited links.

image

### High Contrast White in Firefox

Problems occur when the background is white because focused checkmarks and dots, and expanded triangles, are hardcoded to be white, and they disappear on white background.

image

### High Contrast White in Edge

Same problem in Edge as in Firefox, where focused dots and checkmarks, and expanded menuitem triangles, disappear on white background.

image

### Solutions?

I only know of 2 ways to ensure that an icon will always be visible no matter what High Contrast mode is chosen (other than using [Media Queries, which it is probably best not to rely on for now](https://sarahmhigley.com/writing/whcm-quick-tips/#4.-media-queries)):
1. Draw a border around the icon in a contrasting color so that either the border or the inner part of the icon will always contrast.
2. Use _inline_ SVGs with CSS currentColor so that the icon will always be the same color as the text. It has to be an inline SVG, though, because CSS colors don't seem to want to cross the "document border" into SVG files. At least, I couldn't find a way to get SVG _files_ to follow the currentColor. I also couldn't figure out a way to get the _inline_ svgs to be the nice blue color with WHCM off, and currentColor with WHCM on (without using media queries). @smhigley, Do you know of a way?

**Comment by JAWS-test** [comment]
> I only know of 2 ways to ensure that an icon will always be visible no matter what High Contrast mode is chosen

In my opinion, a third possibility is to use font icons or Unicode characters that are always displayed in the current text color, e.g. `&#8226;` for the dot and `&#9656;` for the triangle

**Comment by jongund** [comment]
@carmacleod 
Thank you for the detailed testing and response, I will switch the example to use inline SVG.

**Comment by carmacleod** [comment]
Thank-you, @jongund.

Please note that the css for the svgs has to have `fill: currentColor;` in order for it to use the current foreground color (i.e. the color of the text).

Sadly, it means that they will always just be black (or white), and that nice blue color will be gone. 😞 
I don't know of a way around that (other than drawing a white border around a blue icon in the svg, or using a media query).
There may be a way - I just don't know what it is.

File: examples/menubar/menubar-navigation.html [hunk 1]
```diff
@@ -95,7 +109,7 @@ <h2 id="ex_label">Example</h2>
                 <a role="menuitem" href="mb-admissions.html#apply">Apply</a>
               </li>
               <li role="none">
-                <a role="menuitem" href="#" aria-haspopup="true" aria-expanded="false">Tuition</a>
+                <a role="menuitem" href="" aria-haspopup="true" aria-expanded="false">Tuition</a>
                 <ul role="menu" aria-label="Tuition Information">
```
Comment (carmacleod) [inline examples/menubar/menubar-navigation.html:113]: For consistency, the menu's aria-label should be the same as the menuitem text.
i.e. After opening the "Tuition" menuitem, I expected to hear "Tuition menu", but instead heard "Tuition Information menu".

```suggestion
                <ul role="menu" aria-label="Tuition">
```

**Comment by jongund** [comment]
@carmacleod 
Can you retest, I updated to use inline SVG and fixed the labeling issue too.

File: examples/menubar/menubar-navigation.html [hunk 2]
```diff
@@ -200,8 +200,12 @@ <h2 id="ex_label">Example</h2>
     <h2>Accessibility Features</h2>
     <ol>
             <li>To help communicate that the arrow keys are available for directional navigation within the menubar and its submenus, a border is added to the menubar container when focus is within the menubar.</li>
-      <li>To support operating system high contrast settings, focus is highlighted by adding and removing a border around the menu item with focus.</li>
-      <li>The down arrow and checked icons are made compatible with operating system high contrast settings and hidden from screen readers by using the CSS <code>content</code> property to render images.</li>
+      <li>To support operating system high contrast settings:
+        <ul>
+          <li>Focus is highlighted by adding and removing a border around the menu item with focus.</li>
+          <li>The arrow icons used to indicate the current state of pull expandable menus use inline SVG images.</li>
```
Comment (carmacleod) [inline examples/menubar/menubar-navigation.html:206]: Probably should also mention currentColor? It's an important part of the support.
(also, probably just say "expandable" and not "pull expandable")

```suggestion
          <li>The arrow icons used to indicate the current state of expandable menus use inline SVG images with CSS currentColor.</li>
```

**Comment by carmacleod** [comment]
Looks great, @jongund !

The triangle icons use the bright yellow link color - very easy to see. I guess currentColor is link color because the menuitems are links.

Here's a screen snap of Navigation Menubar Example in Edge High Contrast Black:

image

And here's a  screen snap of Navigation Menubar Example in Edge High Contrast White:

image

Now just need to fix Editor Menubar Example the same way, but that can be done in a separate PR.

**Comment by jongund** [comment]
@carmacleod 
Thank you the quick review, if all looks good now could you give an approving review.

**Comment by carmacleod** [comment]
@jongund Thanks for updating the Accessibility Features documentation. Please see [my suggestion re talking about currentColor](https://github.com/w3c/aria-practices/pull/1359#discussion_r455933937).
I want to finish testing with screen readers before I can give an approving review. I will get to that later today.

**Comment by carmacleod** [comment]
@jongund Just realized that the 2 links in [aria-practices.html Menubar pattern](https://w3c.github.io/aria-practices/#examples-7) (lines 1938 and 1939 in the code) need to be updated.

**Comment by carmacleod** [comment]
@jongund 
Thanks for fixing the horizontal jitter in the menubar!
If I split the border css into top, bottom, left, right, and give top and bottom the old 4px value, and left and right the new 3px value, then the vertical jitter is gone from the pull down menus as well.

Here's some code that works (this can probably be written more concisely... I'm not a css dev).

... starting line 113 in menubar-navigation.css
```css
  border-top: 4px solid #034575;
  border-bottom: 4px solid #034575;
  border-left: 3px solid #034575;
  border-right: 3px solid #034575;
```

**Comment by jongund** [comment]
@carmacleod 
I think I fixed the gutter issue.
Could you test it when you have a chance.
Thanks again for all your help on this example!

**Comment by mcking65** [comment]
@jongund thank you for all your work on these menubar examples ... awesome improvements!!

And thank you to everyone else who has contributed with reviews and testing. Fabulous teamwork!

Here goes the merge!

**Comment by carmacleod** [comment]
@jongund 
> @carmacleod
> I think I fixed the gutter issue.

Yes!! It's fixed beautifully!  I tested the merged [Navigation Menubar Example](https://w3c.github.io/aria-practices/examples/menubar/menubar-navigation.html) in Chrome, Edge, and Firefox.
No jiggles or jitters anywhere that I could see. :)

Thanks again so much @jongund for all of your work on this!

**Pull request merged**